src/box/txn_limbo.c

```diff
@@ -575,9 +579,11 @@ txn_limbo_read_demote(struct txn_limbo *limbo, int64_t lsn)
 void
 txn_limbo_ack(struct txn_limbo *limbo, uint32_t replica_id, int64_t lsn)
 {
-	assert(!txn_limbo_is_ro(limbo));
 	if (rlist_empty(&limbo->queue))
```

Comment:
What if I collect quorum of ACKs but never had a healthy quorum full? For example, nodes were connecting one by one, ack the change, disconnect. This way I gather a quorum. Looks like I could write a CONFIRM just fine then, but here I won't?
I realized there might be a solution which would solve both this and the problem of async transactions not working even with the empty limbo - drop freeze flag. Then add `replication_synchro_timeout` to the struct limbo as a new field `limbo.timeout`. Instead of a freeze you just set `limbo.timeout` to infinity. Instead of unfreeze you set it to `replication_synchro_timeout`. Also on each update of `replication_synchro_timeout` you set `limbo.timeout` to the same value. It would eliminate this 'frozen' state from the limbo which is complicated enough without it. Will it work?

If you intentionally want to make the entire instance RO, the timeout hack above still would work I think. You can have this frozen flag like you do now, but use it only in `limbo_is_ro()`. The rest of the code here won't change at all from how it was before this patch.

P.S. Why do we forbid ROLLBACKs by the way? Shouldn't we just ban new synchro txns and let the existing transactions end with the regular synchro timeout?

Reply:
I've answered about forbidding rollbacks in previous comment. As for the ACKs collected while limbo is frozen I suggest following changes:
<details>
<summary>diff</summary>

```diff
--- a/src/box/txn_limbo.c
+++ b/src/box/txn_limbo.c
@@ -581,9 +581,6 @@ txn_limbo_ack(struct txn_limbo *limbo, uint32_t replica_id, int64_t lsn)
@@ -581,9 +581,6 @@ txn_limbo_ack(struct txn_limbo *limbo, uint32_t replica_id, int64_t lsn)
 {
        if (rlist_empty(&limbo->queue))
                return;
-       if (limbo->frozen)
-               return;
-       assert(!txn_limbo_is_ro(limbo));
        /*
         * If limbo is currently writing a rollback, it means that the whole
         * queue will be rolled back. Because rollback is written only for
@@ -596,7 +593,7 @@ txn_limbo_ack(struct txn_limbo *limbo, uint32_t replica_id, int64_t lsn)
         */
        if (limbo->is_in_rollback)
                return;
-       assert(limbo->owner_id != REPLICA_ID_NIL);
+       assert(limbo->owner_id == instance_id);
        int64_t prev_lsn = vclock_get(&limbo->vclock, replica_id);
        /*
         * One of the reasons why can happen - the remote instance is not
@@ -631,6 +628,9 @@ txn_limbo_ack(struct txn_limbo *limbo, uint32_t replica_id, int64_t lsn)
        }
        if (confirm_lsn == -1 || confirm_lsn <= limbo->confirmed_lsn)
                return;
+       if (limbo->frozen)
+               return;
+       assert(!txn_limbo_is_ro(limbo));
        txn_limbo_write_confirm(limbo, confirm_lsn);
        txn_limbo_read_confirm(limbo, confirm_lsn);
 }
@@ -927,7 +927,9 @@ txn_limbo_on_parameters_change(struct txn_limbo *limbo)
                        assert(confirm_lsn > 0);
                }
        }
-       if (confirm_lsn > limbo->confirmed_lsn && !limbo->is_in_rollback) {
+       if (confirm_lsn > limbo->confirmed_lsn &&
+           !limbo->is_in_rollback &&
+           limbo->owner_id == instance_id) {
                txn_limbo_write_confirm(limbo, confirm_lsn);
                txn_limbo_read_confirm(limbo, confirm_lsn);
        }
@@ -953,6 +955,7 @@ txn_limbo_unfreeze(struct txn_limbo *limbo)
 {
        limbo->frozen = false;
        box_update_ro_summary();
+       txn_limbo_on_parameters_change(limbo);
 }

 void
```
</details>

That way will still collect ACKs, but won't confirm until we unfreeze our limbo while being a leader.

Reply:
If you managed to deliver your txn to a quorum of nodes, regardless of how, then the new leader also will confirm these txns. But anyway, I don't have a strong opinion here. I will wait until people start using it and come with requests and/or complaints.